# 01. Exploração Inicial dos Dados (EDA)

Objetivo: compreender a estrutura da base Agrofit, avaliar a qualidade dos dados e identificar os principais pontos de atenção para as etapas de limpeza e enriquecimento.

In [1]:
import pandas as pd

CAMINHO_ARQUIVO = "../data/bruto/agrofitprodutosformulados.csv"

df = pd.read_csv(CAMINHO_ARQUIVO, sep=";", encoding="latin1")
print("Base carregada com sucesso.")

Base carregada com sucesso.


## 1. Visão geral da base

Nesta etapa, avalio o tamanho da base, primeiras linhas e colunas disponíveis para entender sua estrutura inicial.

In [2]:
print("Shape da base:", df.shape)

display(df.head())

print("\nColunas:")
print(df.columns.tolist())

Shape da base: (267188, 15)


,NR_REGISTRO,MARCA_COMERCIAL,FORMULACAO,INGREDIENTE_ATIVO,TITULAR_DE_REGISTRO,CLASSE,MODO_DE_ACAO,CULTURA,PRAGA_NOME_CIENTIFICO,PRAGA_NOME_COMUM,EMPRESA_PAIS_TIPO,CLASSE_TOXICOLOGICA,CLASSE_AMBIENTAL,ORGANICOS,SITUACAO
0,35523,KBR-829M1-02,Nematóides vivos,Heterorhabditis bacteriophora (Nematóides ent...,Koppert do Brasil Holding S.A. - Piracicaba/SP,Agente Biológico de Controle,NaN,Todas as culturas,Scaptocoris castanea,Percevejo-castanho,(Koppert do Brasil Holding S.A. - Piracicaba/...,Não Classificado - Produto Não Classificado,Produto Pouco Perigoso ao Meio Ambiente,NAO,True
1,35523,KBR-829M1-02,Nematóides vivos,Heterorhabditis bacteriophora (Nematóides ent...,Koppert do Brasil Holding S.A. - Piracicaba/SP,Agente Biológico de Controle,NaN,Todas as culturas,Sphenophorus levis,Bicudo da cana-de-açúcar; Gorgulho-da-cana,(Koppert do Brasil Holding S.A. - Piracicaba/...,Não Classificado - Produto Não Classificado,Produto Pouco Perigoso ao Meio Ambiente,NAO,True
2,35523,KBR-829M1-02,Nematóides vivos,Heterorhabditis bacteriophora (Nematóides ent...,Koppert do Brasil Holding S.A. - Piracicaba/SP,Agente Biológico de Controle,NaN,Todas as culturas,Spodoptera frugiperda,Lagarta-militar,(Koppert do Brasil Holding S.A. - Piracicaba/...,Não Classificado - Produto Não Classificado,Produto Pouco Perigoso ao Meio Ambiente,NAO,True
3,33723,KBR-S39M1-02,Nematóides vivos,Steinernema carpocapsae (Nematóides entomopat...,Koppert do Brasil Holding S.A. - Piracicaba/SP,Agente Biológico de Controle,NaN,Todas as culturas,Bradysia matogrossensis,fungus gnats,(Koppert do Brasil Holding S.A. - Piracicaba/...,Não Classificado - Produto Não Classificado,Produto Pouco Perigoso ao Meio Ambiente,NAO,True
4,33723,KBR-S39M1-02,Nematóides vivos,Steinernema carpocapsae (Nematóides entomopat...,Koppert do Brasil Holding S.A. - Piracicaba/SP,Agente Biológico de Controle,NaN,Todas as culturas,Sphenophorus levis,Bicudo da cana-de-açúcar; Gorgulho-da-cana,(Koppert do Brasil Holding S.A. - Piracicaba/...,Não Classificado - Produto Não Classificado,Produto Pouco Perigoso ao Meio Ambiente,NAO,True



Colunas:
['NR_REGISTRO', 'MARCA_COMERCIAL', 'FORMULACAO', 'INGREDIENTE_ATIVO', 'TITULAR_DE_REGISTRO', 'CLASSE', 'MODO_DE_ACAO', 'CULTURA', 'PRAGA_NOME_CIENTIFICO', 'PRAGA_NOME_COMUM', 'EMPRESA_PAIS_TIPO', 'CLASSE_TOXICOLOGICA', 'CLASSE_AMBIENTAL', 'ORGANICOS', 'SITUACAO']


## 2. Estrutura dos dados

Aqui verifico tipos de dados e preenchimento das colunas.

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 267188 entries, 0 to 267187
Data columns (total 15 columns):
 #   Column                 Non-Null Count   Dtype
---  ------                 --------------   -----
 0   NR_REGISTRO            267188 non-null  int64
 1   MARCA_COMERCIAL        267188 non-null  str  
 2   FORMULACAO             267172 non-null  str  
 3   INGREDIENTE_ATIVO      267188 non-null  str  
 4   TITULAR_DE_REGISTRO    267188 non-null  str  
 5   CLASSE                 267188 non-null  str  
 6   MODO_DE_ACAO           152352 non-null  str  
 7   CULTURA                267157 non-null  str  
 8   PRAGA_NOME_CIENTIFICO  266117 non-null  str  
 9   PRAGA_NOME_COMUM       264163 non-null  str  
 10  EMPRESA_PAIS_TIPO      267092 non-null  str  
 11  CLASSE_TOXICOLOGICA    267188 non-null  str  
 12  CLASSE_AMBIENTAL       267188 non-null  str  
 13  ORGANICOS              267188 non-null  str  
 14  SITUACAO               267188 non-null  bool 
dtypes: bool(1), int64(1), str(13

## 3. Análise de valores nulos

Identifico colunas com dados ausentes e seu impacto percentual na base.

In [4]:
nulos = df.isnull().sum().sort_values(ascending=False)
percentual_nulos = ((df.isnull().sum() / len(df)) * 100).sort_values(ascending=False)

display(nulos)
display(percentual_nulos.round(2))

MODO_DE_ACAO             114836
PRAGA_NOME_COMUM           3025
PRAGA_NOME_CIENTIFICO      1071
EMPRESA_PAIS_TIPO            96
CULTURA                      31
FORMULACAO                   16
NR_REGISTRO                   0
CLASSE                        0
TITULAR_DE_REGISTRO           0
INGREDIENTE_ATIVO             0
MARCA_COMERCIAL               0
CLASSE_TOXICOLOGICA           0
CLASSE_AMBIENTAL              0
ORGANICOS                     0
SITUACAO                      0
dtype: int64

MODO_DE_ACAO             42.98
PRAGA_NOME_COMUM          1.13
PRAGA_NOME_CIENTIFICO     0.40
EMPRESA_PAIS_TIPO         0.04
CULTURA                   0.01
FORMULACAO                0.01
NR_REGISTRO               0.00
CLASSE                    0.00
TITULAR_DE_REGISTRO       0.00
INGREDIENTE_ATIVO         0.00
MARCA_COMERCIAL           0.00
CLASSE_TOXICOLOGICA       0.00
CLASSE_AMBIENTAL          0.00
ORGANICOS                 0.00
SITUACAO                  0.00
dtype: float64

## 4. Verificação de duplicidade

Avalio se existem registros duplicados na base.

In [5]:
print("Duplicados exatos:", df.duplicated().sum())

Duplicados exatos: 0


## 5. Granularidade dos dados

Avalio se a base está no nível de produto ou se há múltiplos registros por produto.

In [6]:
produtos_unicos = df["NR_REGISTRO"].nunique()

print("Total de linhas:", len(df))
print("Produtos únicos:", produtos_unicos)
print("Média de linhas por produto:", round(len(df) / produtos_unicos, 2))

Total de linhas: 267188
Produtos únicos: 4238
Média de linhas por produto: 63.05


## 6. Cardinalidade das variáveis

Analiso a quantidade de valores únicos em colunas relevantes para entender sua diversidade.

In [7]:
colunas_relevantes = [
    "MARCA_COMERCIAL",
    "FORMULACAO",
    "INGREDIENTE_ATIVO",
    "TITULAR_DE_REGISTRO",
    "CLASSE",
    "MODO_DE_ACAO",
    "CULTURA",
    "PRAGA_NOME_CIENTIFICO",
    "PRAGA_NOME_COMUM",
    "EMPRESA_PAIS_TIPO",
    "CLASSE_TOXICOLOGICA",
    "CLASSE_AMBIENTAL",
    "ORGANICOS",
    "SITUACAO"
]

cardinalidade = pd.DataFrame({
    "coluna": colunas_relevantes,
    "valores_unicos": [df[col].nunique(dropna=True) for col in colunas_relevantes]
})

display(cardinalidade.sort_values("valores_unicos", ascending=False))

,coluna,valores_unicos
0,MARCA_COMERCIAL,4238
9,EMPRESA_PAIS_TIPO,3282
2,INGREDIENTE_ATIVO,1737
8,PRAGA_NOME_COMUM,1710
7,PRAGA_NOME_CIENTIFICO,1455
5,MODO_DE_ACAO,452
3,TITULAR_DE_REGISTRO,343
6,CULTURA,238
4,CLASSE,50
1,FORMULACAO,50


## 7. Distribuições iniciais

Visualizo as categorias mais frequentes em algumas variáveis-chave.

In [8]:
print("Top 10 empresas:")
display(df["TITULAR_DE_REGISTRO"].value_counts().head(10))

print("\nTop 10 classes:")
display(df["CLASSE"].value_counts().head(10))

print("\nTop 10 culturas:")
display(df["CULTURA"].value_counts().head(10))

print("\nTop 10 ingredientes ativos:")
display(df["INGREDIENTE_ATIVO"].value_counts().head(10))

Top 10 empresas:


TITULAR_DE_REGISTRO
AllierBrasil Agro Ltda.                                                                       18431
Nortox S.A. - Arapongas                                                                       17788
Rainbow Defensivos Agrícolas Ltda.- Porto Alegre /RS                                         12411
Sinon do Brasil Ltda. - Porto Alegre /RS.                                                     10056
Syngenta Proteção de Cultivos Ltda. – São Paulo/SP                                         9708
Wynca do Brasil Ltda – São Paulo                                                             9644
Sumitomo Chemical Brasil Indústria Química S.A. - Maracanaú/CE                              7371
UPL do Brasil Indústria e Comércio de Insumos Agropecuários S.A. - Matriz Ituverava/SP      7117
CHDS do Brasil Comércio de Insumos Agrícolas Ltda - São Miguel do Iguaçu/PR                7080
Cropchem Ltda - Porto Alegre                                                    


Top 10 classes:


CLASSE
Herbicida                             212870
Fungicida                              19439
Inseticida                             16462
Acaricida/Inseticida                    6032
Acaricida/Fungicida                     2934
Inseticida Microbiológico              1200
Herbicida/Regulador de Crescimento       944
Bactericida/Fungicida                    748
Acaricida                                537
Fungicida microbiológico                527
Name: count, dtype: int64


Top 10 culturas:


CULTURA
Soja                20999
Milho               15148
Cana-de-açúcar    15020
Café               12320
Algodão            12044
Citros              11711
Trigo               11186
Pastagens            9007
Maçã               8710
Eucalipto            8564
Name: count, dtype: int64


Top 10 ingredientes ativos:


INGREDIENTE_ATIVO
glifosato-sal de isopropilamina (glicina substituída) (480 g/L)     55466
glifosato-sal de amônio (glicina substituída) (792.5 g/kg)         23462
glifosato (glicina substituída) (480 g/L)                           17169
Glufosinato - sal de amônio (homoalanina substituída) (200 g/L)    11418
cletodim (oxima ciclohexanodiona) (240 g/L)                           9446
glifosato (glicina substituída) (792.5 g/kg)                         6303
glifosato-sal de amônio (glicina substituída) (757 g/kg)            4244
glifosato-sal de isopropilamina (glicina substituída) (648 g/L)      3440
glifosato-sal de amônio (glicina substituída) (792.5 g/L)           3326
glifosato (glicina substituída) (356 g/L)                            2709
Name: count, dtype: int64

## 8. Principais achados

A partir da exploração inicial, foram identificados os seguintes pontos:

- A base possui volume relevante (mais de 260 mil registros).
- Não foram identificadas duplicatas exatas.
- A coluna `MODO_DE_ACAO` apresenta alto volume de valores nulos.
- Há indícios de problemas de encoding em colunas textuais.
- A granularidade da base não é por produto, mas sim produto associado a múltiplas pragas/culturas.
- A coluna `EMPRESA_PAIS_TIPO` aparenta conter informações compostas (empresa + localização).
- A base possui alta diversidade em ingredientes ativos e pragas, indicando potencial analítico relevante.

## Diagnóstico geral da base

A base Agrofit apresenta boa completude na maioria das colunas, com ausência relevante apenas na variável `MODO_DE_ACAO`, que possui cerca de 43% de valores nulos.

A análise de granularidade evidenciou que os registros não estão no nível de produto, mas sim associados a múltiplas pragas e culturas, resultando em uma média de 63 registros por produto.

Dessa forma, será necessário estruturar a base de forma analítica, evitando contagens duplicadas em análises agregadas.